xarray/core/rolling_exp.py

```diff
@@ -128,9 +129,41 @@ def mean(self, keep_attrs: bool | None = None) -> T_DataWithCoords:
         if keep_attrs is None:
             keep_attrs = _get_keep_attrs(default=True)
 
-        return self.obj.reduce(
-            move_exp_nanmean, dim=self.dim, alpha=self.alpha, keep_attrs=keep_attrs
-        )
+        dim_order = self.obj.dims
+
+        from xarray.core.dataset import Dataset
+
+        if isinstance(self.obj, Dataset):
+            new_ds = self.obj.copy()
+            if not keep_attrs:
+                new_ds.attrs = {}
+            for name, var in self.obj.variables.items():
+                if self.dim in self.obj[name].dims:
+                    new_ds[name] = apply_ufunc(
+                        move_exp_nanmean,
+                        var,
+                        input_core_dims=[[self.dim]],
+                        kwargs=dict(alpha=self.alpha, axis=-1),
```

Comment:
`reduce` doe the useful thing of [making the `dim` -> `axis` translation](https://github.com/pydata/xarray/blob/ab096b05821a039f7a5b416e83227462625687ff/xarray/core/variable.py#L1991), rather than transposing, like `apply_ufunc` does. 

Though if `apply_ufunc` restored dimension order, it would basically equivalent

Reply:
Doesn't numpy require the core dim to be at the end (or some such, I could be wrong, though).

Reply:
I think this used to be a constraint, but more recently we can pass `axes` or `axis` to gufuncs: https://numpy.org/doc/stable/reference/ufuncs.html. I made a change in numbagg to take advantage of this: https://github.com/numbagg/numbagg/pull/94

(though possibly I'm misunderstanding, and possibly there are functions that still require axis moving?)